specs/agents/tracing-api-otel.md

```diff
@@ -0,0 +1,251 @@
+## OpenTelemetry API (Tracing)
+
+[OpenTelemetry](https://opentelemetry.io) (OTel in short) provides a vendor-neutral API that allows to capture tracing, logs and metrics data.
+
+Agents MAY provide a bridge implementation of OpenTelemetry Tracing API following this specification.
+When available, implementation MUST be configurable and should be disabled by default when marked as `experimental`.
+
+Bridging here means that for each OTel span created with the API, a native span/transaction will be created and sent to APM server.
+
+From the perspective of the application code calling the OTel API, the delegation to a native span/transaction is transparent.
+Also, this means that any OTel processors will be bypassed and ignored by the bridge.
+
+### Spans and Transactions
+
+OTel only defines Spans, whereas Elastic APM relies on both Spans and Transactions.
+OTel allows users to provide the _remote context_ when creating a span, which is equivalent to providing a parent to a transaction or span,
+it also allows to provide a (local) parent span.
+
+As a result, when creating Spans through OTel API with a bridge, agents must implement the following algorithm:
+
+```javascript
+// otel_span contains the properties set through the OTel API
+span_or_transaction = null;
+if (otel_span.remote_contex != null) {
+    span_or_transaction = createTransactionWithParent(otel_span.remote_context);
+} else if (otel_span.parent == null) {
+    span_or_transaction = createRootTransaction();
+} else {
+    span_or_transaction = createSpanWithParent(otel_span.parent);
+}
+```
+
+### Span Kind
+
+OTel spans have an `SpanKind` property ([specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)) which is close but not strictly equivalent to our definition of spans and transactions.
+
+For both transactions and spans, an optional `otel.span_kind` property will be provided by agents when set through
+the OTel API.
+This value should be stored into Elasticsearch documents to preserve OTel semantics and help future OTel integration.
+
+Possible values are `CLIENT`, `SERVER`, `PRODUCER`, `CONSUMER` and `INTERNAL`, refer to [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind) for details on semantics.
+
+When `otel.span_kind` is not provided by the agent, APM Server should infer it using the following algorithm:
```

Comment:
Is this a common case in other agents? The [python sdk](https://github.com/open-telemetry/opentelemetry-python/blob/44e9b5531d91c8c52f58cd4ea3930cda5c1ed662/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L972) sets `kind = SpanKind.INTERNAL` by default, I don't think you can have an empty SpanKind when using the main Tracer. I think python will always have `otel.span_kind` set.

Reply:
The description is a bit misleading. This is about the case when the agent sends regular data that doesn't come from the OTel bridge. In that case, `otel.span_kind` is empty. To ensure that there's always a `span_kind`, no matter whether the agents send it, APM Server auto-infers the `span_kind`. There's nothing the agents need to do.

Reply:
Oh! Then maybe we could use `span_kind` for telemetry purposes. See my previous comment.

Reply:
I will clarify this statement to make it clear that:
- when using the OTel bridge, the `span_kind` is set to `SpanKind.INTERNAL` by default by the OTel API implementation.
- when the agent produces a span/transaction that does not originates from the OTel bridge and no explicit `otel.span_kind` have been set, then we have to infer it from span/transaction data in order to populate the `otel.span_kind` field that was added in apm-server 7.16.